[exporter/awss3] add marsheler config to pick only body content of logs

anthoai97 opened the pull request:
**Description:** <Describe what has changed.>
<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.-->
This pull request introduces a new marshaller for the awss3 exporter. It disregards resources, labels, and selectively extracts only the body content of the logs, facilitating simplified AWS Athena connection and querying. While straightforward, this enhancement accommodates various use cases, enabling the extraction of specific logs for statistical analysis and long-term archival reporting.

Example when write only body to S3:
```
EXTERNAL_REQUEST 2023-07-18T10:05:01.976Z GET 200 localhost /organization/v0/properties 0:0:0:0:0:0:0:1 4264 348 wUN0NeTMByHVmIOjo79ImwylAjjkRtkr
EXTERNAL_REQUEST 2023-07-18T10:05:01.976Z GET 200 localhost /organization/v0/properties 0:0:0:0:0:0:0:1 4264 348 wUN0NeTMByHVmIOjo79ImwylAjjkRtkr
EXTERNAL_REQUEST 2023-07-18T10:05:01.976Z GET 200 localhost /organization/v0/properties 0:0:0:0:0:0:0:1 4264 348 wUN0NeTMByHVmIOjo79ImwylAjjkRtkr
EXTERNAL_REQUEST 2023-07-18T10:05:01.976Z GET 200 localhost /organization/v0/properties 0:0:0:0:0:0:0:1 4264 348 wUN0NeTMByHVmIOjo79ImwylAjjkRtkr
```

**Testing:** <Describe what testing was performed and which tests were added.>
Work as expected when used Athena to Query:
<img width="962" alt="athena" src="https://github.com/open-telemetry/opentelemetry-collector-contrib/assets/85854989/8f682024-df4e-402b-9de1-ee8268ea1936">

**Documentation:** <Describe the documentation added.>
Readme updated.

atoulme commented:
Please rebase

anthoai97 commented:
> Please rebase

@atoulme i'm rebased, thank you !!!

atoulme commented on exporter/awss3exporter/README.md:
```diff
@@ -42,6 +42,8 @@ Marshaler determines the format of data sent to AWS S3. Currently, the following
 - `otlp_json` (default): the [OpenTelemetry Protocol format](https://github.com/open-telemetry/opentelemetry-proto), represented as json.
 - `sumo_ic`: the [Sumo Logic Installed Collector Archive format](https://help.sumologic.com/docs/manage/data-archiving/archive/).
   **This format is supported only for logs.**
+- `body_log`: In case you only want to wrap logs body for simple to use Athena query.
```
Is Athena the only use case?

anthoai97 replied:
No, it's not the only case. There are numerous use cases where focusing solely on the body content can be beneficial:
- This approach helps minimize the transformation cost from JSON to aggregated relational data, facilitating the creation of dashboards. 
- when combined with a body-filtering processor, it becomes conducive to archiving data for long-term use. 
- By using logs as measure data, this strategy also contributes to a reduction in the main system database write I/O.

BTW: I updated the document to make it more general.

atoulme commented on exporter/awss3exporter/README.md:
```diff
@@ -42,6 +42,8 @@ Marshaler determines the format of data sent to AWS S3. Currently, the following
 - `otlp_json` (default): the [OpenTelemetry Protocol format](https://github.com/open-telemetry/opentelemetry-proto), represented as json.
 - `sumo_ic`: the [Sumo Logic Installed Collector Archive format](https://help.sumologic.com/docs/manage/data-archiving/archive/).
   **This format is supported only for logs.**
+- `body_log`: If you specifically intended to export only the body of logs and ignore other labels.
```
```suggestion
- `body_log`: export the log body as string
```
Can we change from `body_log` to `body`?

anthoai97 replied:
Yes, I can. It's changed to "body" in the latest commit.

atoulme commented on exporter/awss3exporter/logbody_marshaler.go:
```diff
@@ -0,0 +1,52 @@
+type bodylogMarshaler struct{}
+
+func (*bodylogMarshaler) format() string {
+	return "log"
```
should it be "log"?

anthoai97 replied:
I modified the format to `.txt` for consistency with the document that states _'export the log body as a string.'_

atoulme commented on exporter/awss3exporter/logbody_marshaler.go:
```diff
@@ -0,0 +1,52 @@
+			for k := 0; k < logs.Len(); k++ {
+				lr := logs.At(k)
+				body := lr.Body()
+				buf.WriteString(body.Str())
```
this doesn't work well for values that are not of type string. Please write a few test cases with a boolean, int, or map body. Use `AsString()` instead of `Str()`.

anthoai97 replied:
Thank you for sharing your insights. I have updated to `AsString()` and added the test cases to cover various data types, including boolean, number, and map. Please kindly review my latest commit.

atoulme commented on .chloggen/feat_allow_pick_only_log_body.yaml:
```diff
@@ -0,0 +1,26 @@
+component: awss3exporter
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: Add config option is only pick the body of logs and ignore other labels
```
```suggestion
note: Add a marshaler that stores the body of log records in s3.
```

atoulme commented:
Thank you, this looks good. A last nit on the changelog note and I am ready to accept the PR.

anthoai97 commented:
> Thank you, this looks good. A last nit on the changelog note and I am ready to accept the PR.

I'm pleased to hear that. I updated the changelog and changed the variable name to pass the lint workflow. Thank you.

anthoai97 commented:
Hi @atoulme, could you please guide the steps I need to take to merge into the 'main' branch? Are there any specific requirements or actions I should be aware of?

atoulme commented:
Sorry, please rebase and I can flag for maintainers to take a look.

anthoai97 committed:
[exporter/awss3] add marsheler config to pick only body content of logs

anthoai97 commented:
Dear @atoulme, I have rebased the branch, but unfortunately, I made a mistake during the process which added many reviewers to the review list. Could you please advise me on how to resolve this issue?

Thank you very much !!!
